### PR TITLE
Match condition element order to body element order

### PR DESCRIPTION
**Pull request by morettid**
Solves issue #83.

### CODE REVIEW

File: tests/test_kratos_model_io.py [hunk 1]
```diff
@@ -182,4 +244,32 @@ def test_write_mdpa_text(self, create_default_2d_model_and_mesh):
 
         expected_text_body = [line.rstrip() for line in expected_text_body]
         # assert the objects to be equal
-        npt.assert_equal(actual=actual_text_body, desired=expected_text_body)
\ No newline at end of file
+        npt.assert_equal(actual=actual_text_body, desired=expected_text_body)
+
+    def test_write_mdpa_text_2_layers(self, create_default_2d_2_layers_model_and_mesh):
```
Comment (aronnoordam) [inline tests/test_kratos_model_io.py:249]: type hint missing

File: stem/model.py [hunk 2]
```diff
@@ -326,6 +327,135 @@ def generate_mesh(self):
...
+            #  from a 2D quadratic beam (3 nodes) , it should return 2.
+            #  from a 2D triangular quadratic element (6 nodes) it should return 3.
+            n_nodes = len(nodes_process_element)
+            # append the first (n_nodes-1) elements at the back of to the list (to make a cycle)
+            target_list = nodes_body_element + nodes_body_element[:(n_nodes-1)]
+            # pick only the corner nodes, the mid-point nodes related to the higher order do not matter.
+            source_list = nodes_process_element[:n_nodes]
+
+            if not Utils.has_matching_combination(target_list, source_list):
+                process_element.node_ids = nodes_process_element[::-1]
```
Comment (aronnoordam) [inline stem/model.py:457]: this is never tests

File: tests/test_utils.py [hunk 3]
```diff
@@ -143,6 +143,21 @@ def test_chain_sequence(self):
...
+        sub_list2 = [5, 1]
+        sub_list3 = [1, 3, 4]
+        sub_list4 = [1, 3, 5]
+
+        list_tst = [1, 3, 4, 5, 1]
+        assert not Utils.has_matching_combination(list_tst, sub_list1)
+        assert Utils.has_matching_combination(list_tst, sub_list2)
+        assert Utils.has_matching_combination(list_tst, sub_list3)
+        assert not Utils.has_matching_combination(list_tst, sub_list4)
+
```
Comment (aronnoordam) [inline tests/test_utils.py:160]: can you add extra tests which check if the errors are raised correctly

File: stem/mesh.py [hunk 4]
```diff
@@ -151,12 +153,22 @@ def create_mesh_from_gmsh_group(cls, mesh_data: Dict[str, Any], group_name: str)
...
+        nodes: Dict[int, Node] = {node_id: Node(node_id, mesh_data["nodes"][node_id]) for node_id in group_node_ids}
+
+        # add each element, but first check if counterclockwise
+        # revert the node id order if it is not
+        elements: Dict[int, Element] = {}
+        # create element per element id
+        for element_id in group_element_ids:
+
+            node_ids_element = element_type_data[element_id]
+            # flip the element nodes if they are not anti-clockwise, and only if mesh entity has more than 2 nodes.
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply stem/mesh.py:166]: i think this functionality (reordering anti-clockwise) deserves its own function

File: stem/model.py [hunk 5]
```diff
@@ -326,6 +327,135 @@ def generate_mesh(self):
...
+
+            nodes_process_element = [_id for _id in process_element.node_ids]
+            nodes_body_element = [_id for _id in body_element.node_ids]
+            # TODO: a function to get the number of nodes of the elements nevertheless the integration order.
+            #  from a 2D quadratic beam (3 nodes) , it should return 2.
+            #  from a 2D triangular quadratic element (6 nodes) it should return 3.
+            n_nodes = len(nodes_process_element)
+            # append the first (n_nodes-1) elements at the back of to the list (to make a cycle)
+            target_list = nodes_body_element + nodes_body_element[:(n_nodes-1)]
+            # pick only the corner nodes, the mid-point nodes related to the higher order do not matter.
```
[the remark this answers is not in the document]
Comment (morettid) [reply stem/model.py:453]: True, will add a TODO

File: stem/model.py [hunk 6]
```diff
@@ -326,6 +327,135 @@ def generate_mesh(self):
...
+    def __find_matching_body_elements_for_process_model_part(
+            self, process_model_part: ModelPart, check_all_coupled: bool = False
+    ):
+        """
+        For a process model part, tries finds the matching body elements on which the condition elements are applied.
+
+        Args:
+            - process_model_part (:class:`stem.model_part.ModelPart`): model part from which element nodes needs to be
+                extracted.
+            - check_all_coupled (bool): checks if the condition element are applied to a body model part and raises an
```
[the remark this answers is not in the document]
Comment (morettid) [reply stem/model.py:367]: Input can be set to true if we want to enforce the match of all process elements to a body element.
I.e. can be used later to check in 2D that process model parts are applied on body model parts.

In 3D this is not yet available.

**Comment by aronnoordam** [comment]
closed because PR should now go towards the main branch, #102

**Pull request closed without merging**